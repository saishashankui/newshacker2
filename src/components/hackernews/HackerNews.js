import React from 'react';
import { Link } from 'react-router-dom';
import './HackerNewsStyles.css';

class HackerNews extends React.Component {
    constructor(props) {
        super(props);
        this.state = { stories: [] };
    }
    async componentDidMount() {
        try {
            const apiUrl = 'https://hacker-news.firebaseio.com/v0/topstories.json';
            await fetch(apiUrl)
                .then((response) => response.json())
                .then((data) =>
                    data.slice(0, 10).map((story) => (
                        fetch(`https://hacker-news.firebaseio.com/v0/item/${story}.json`)
                            .then((response) => response.json())
                            .then((data) =>
                                this.setState(state => ({
                                    stories: [...state.stories, data]
                                }))
                            )
                    ))
                )
        } catch (err) {
            // console.error(err);
        }
    }
    render() {
        return (
            <div>
                <header className="masthead clear">
                    <div className="centered">
                        <div className="site-branding">
                            <h1 className="site-title">Hacker News</h1>
                            <p>Top 10 Stories</p>
                        </div>
                    </div>
                </header>
                <main className="main-area">
                    <div className="centered">
                        <section className="cards">


                            {this.state.stories.map(function (storyView, index) {
                                return <article key={`${index}_key`} className="card">
                                    <h2>
                                        <Link key={`${index}_link`} to={{ pathname: `story/${storyView.id}`, storyView: storyView }} >
                                            <div className="card-content" key={`${index}_cc`} >{storyView.title}</div>
                                            <p key={`${index}_date`} className="m-0">
                                                {new Date(storyView.time * 1000).toDateString()}
                                            </p>
                                        </Link>
                                    </h2>
                                </article>
                            })}
                        </section>
                    </div>
                </main>
            </div>
        )
    }
}
export default HackerNews;