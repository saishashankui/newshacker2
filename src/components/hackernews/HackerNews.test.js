import { render, screen } from '@testing-library/react';
import { shallow } from "enzyme";
import HackerNews from './HackerNews';

it('renders Hacker news header', () => {
    render(<HackerNews />);
    const linkElement = screen.getByText(/Hacker News/i);
    expect(linkElement).toBeInTheDocument();
});

beforeAll(() => {
    // global.fetch = jest.fn();
    window.fetch = jest.fn(); //if running browser environment
});

let wrapper;
beforeEach(() => {
    wrapper = shallow(<HackerNews />, { disableLifecycleMethods: true });
});
afterEach(() => {
    wrapper.unmount();
});

it("must render a loading span before api call success", () => {
    
    expect(wrapper.find("section.cards").exists()).toBeTruthy();
    expect(wrapper.debug()).toMatchSnapshot();
    const spyDidMount = jest.spyOn(HackerNews.prototype, "componentDidMount");

    fetch.mockImplementation(() => {
        return Promise.resolve({
            status: 200,
            json: () => {
                return Promise.resolve({
                    "id": 26781052,
                    "title": "How People Get Rich Now"
                });
            }
        });
    });
    const didMount = wrapper.instance().componentDidMount();
    // expecting componentDidMount have been called
    expect(spyDidMount).toHaveBeenCalled();
    
    didMount.then(() => {
        // updating the wrapper
        wrapper.update();
        expect(wrapper.find("div.card-content").text()).toContain("How People Get Rich Now");
        spyDidMount.mockRestore();
        fetch.mockClear();
        done();
    });
});