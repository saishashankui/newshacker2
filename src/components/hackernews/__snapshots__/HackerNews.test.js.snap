// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must render a loading span before api call success 1`] = `
"<div>
  <header className=\\"masthead clear\\">
    <div className=\\"centered\\">
      <div className=\\"site-branding\\">
        <h1 className=\\"site-title\\">
          Hacker News
        </h1>
        <p>
          Top 10 Stories
        </p>
      </div>
    </div>
  </header>
  <main className=\\"main-area\\">
    <div className=\\"centered\\">
      <section className=\\"cards\\" />
    </div>
  </main>
</div>"
`;
